<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>Genome Size Constrains with Horizontal Gene Transfer: Ecosystem Class Reference</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">Genome Size Constrains with Horizontal Gene Transfer
   
   </div>
   <div id="projectbrief">by Piotr Bentkowski</div>
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('classEcosystem.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a>  </div>
  <div class="headertitle">
<div class="title">Ecosystem Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<!-- doxytag: class="Ecosystem" -->
<p>This is 'The <a class="el" href="classEcosystem.html" title="This is &#39;The Ecosystem&#39;. A place where things are happening.">Ecosystem</a>'. A place where things are happening.  
 <a href="classEcosystem.html#details">More...</a></p>

<p><code>#include &quot;<a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>&quot;</code></p>

<p><a href="classEcosystem-members.html">List of all members.</a></p>
<table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#aad52d5b02f1e2f0ffd16c72825dee140">Ecosystem</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Constructor - sets an ecosystem, but with no values.  <a href="#aad52d5b02f1e2f0ffd16c72825dee140"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a9c459456082bdcebebeaba67b276cfd0">setPopulation</a> (double total_resource_pool, double Env, double minimal_cells_resource, double maximal_resourse_bonus_at_start, int min_genotype_initialized, int max_genotype_initialized, double not_super_gene_param, double gene_space_width, double environment_resolution, double one_over_total_env_space_surf)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Sets a new population in given Env conditions and of given size.  <a href="#a9c459456082bdcebebeaba67b276cfd0"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#ae1f5162ddcbbec9933d5e2c5c62678fe">reproduceOrKillCell</a> (std::string mode_of_feeding, double Env, double duplication_probability, double deletion_probability, double mutate_threshold, int minimum_genotype_size, double not_super_gene_param, double gene_space_width, double environment_resolution, double one_over_total_env_space_surf, double random_death_rate, double minimal_cells_resource, double reproduction_resourse_size, double disparity_at_reproduction)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Reproduces or/and kills cells.  <a href="#ae1f5162ddcbbec9933d5e2c5c62678fe"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a144a0594e881ad7b36b64faccf1d9b4c">feedCells</a> (std::string mode_of_feeding, double Env, double gene_replication_cost, int core_genes, double metabolic_costs, double gain_from_the_genotype_ratio)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Feeds cells. Has two modes of feeding: (p) proportionally to cell's demand for resource and to resource availability and (s) sequential feeding when cells are aligned in a random queue and fed one by one until depleting resources (not all cells are fed then).  <a href="#a144a0594e881ad7b36b64faccf1d9b4c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#ace4318bc4766978edea2f73ebcaf98f8">updateGrowthRatesOfCellsInPopulation</a> (double Env)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Updates growth rates of all the cells in the ecosystem.  <a href="#ace4318bc4766978edea2f73ebcaf98f8"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a4dccae9a322022d09724501419df349f">horizontalGeneTransfer</a> (double hgt_donor_prob, double hgt_gene_trans)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Core method. Does the horizontal gene transfer.  <a href="#a4dccae9a322022d09724501419df349f"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a6549e81ea4daa38924a3791f649e1833">printNumberOfGenesInCells</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Prints a number of genes in genotypes of all Cells.  <a href="#a6549e81ea4daa38924a3791f649e1833"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#ae4373e033efacb2fc1d8be6b5521880c">returnMeanNumberOfGenesInCells</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the mean number of genes in genotypes of all the cells.  <a href="#ae4373e033efacb2fc1d8be6b5521880c"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a30fc9729432ba63f068f19effa42d8dd">returnStdOfGenesInCells</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the standard deviation of the number of genes in genotypes of all the cells.  <a href="#a30fc9729432ba63f068f19effa42d8dd"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a6e753a4d44bae33695219d15a8324930">returnMeanAgeOfCells</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the mean age of all cells in ecosystem.  <a href="#a6e753a4d44bae33695219d15a8324930"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a59bedd0e20191831596d1f5a8f8f6ad2">returnSTDAgeOfCells</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the standard deviation of the age of cells.  <a href="#a59bedd0e20191831596d1f5a8f8f6ad2"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a5817a0df8a9753d2fb21fe2a465b1364">returnsSummOfCellResources</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns returns the sum of all resources allocated in all the cells.  <a href="#a5817a0df8a9753d2fb21fe2a465b1364"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a29523e18d64fa67c8aebb9a99aff5957">histGenomeSize</a> (int genome_size_histogram_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Writes to file a histogram of sizes of genotypes in given time step.  <a href="#a29523e18d64fa67c8aebb9a99aff5957"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a611ff2fecef4a34dcb73657ad572dbf8">histCellsAge</a> (int histSize)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns a histogram of ages of cells.  <a href="#a611ff2fecef4a34dcb73657ad572dbf8"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#aff0fcbd5b5fbfa21fd40401d35a5a35b">returnPopulationShannonIndex</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the Shannon index value for population diversity. Cells which have even one gene different will be accounted as members of different clonal strains.  <a href="#aff0fcbd5b5fbfa21fd40401d35a5a35b"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a29837c080dcab24bb549b54f726b7836">returnGenesShannonIndex</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the Shannon Index for genes in the population. Each gene, regardless in which cell is located, is treated as an independent entity. Uses <a href="http://www.gnu.org/software/gsl/manual/html_node/Logarithm-and-Related-Functions.html">GNU Scientific Library logarithm function</a>  <a href="#a29837c080dcab24bb549b54f726b7836"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a5853f9762c4c91d40dfb05c959abd097">histGenotypeEnvelopeSurface</a> (int envelope_of_genotype_hist_size, double envelope_of_genotype_hist_range)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns a histogram of fractions of total environmental space taken by genotypes.  <a href="#a5853f9762c4c91d40dfb05c959abd097"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a8ab8b96cd612eebec075a5a98feeec29">histGenotypeEnvelopeMaximum</a> (int maximum_of_genotype_hist_size, double maximum_of_genotype_hist_range)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns a histogram of maximums of genotypes envelopes.  <a href="#a8ab8b96cd612eebec075a5a98feeec29"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#ab658ff43ae86399189569f9b00222c65">histResourceAllocatedInCells</a> (int resource_alloc_hist_size, double resource_alloc_hist_range, double resource_alloc_hist_start)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns a histogram of resources allocated in the cells.  <a href="#ab658ff43ae86399189569f9b00222c65"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#acd7c553337ea8a38ecb4a1572cf68b52">histResourceUptakenByCells</a> (int resource_uptake_hist_size, double resource_uptake_hist_range)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns a histogram of resources allocated in the cells.  <a href="#acd7c553337ea8a38ecb4a1572cf68b52"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a882e846f4004e31b88e801514c932484">histsAgeAndResourceAtReprod</a> (int at_repr_intake_hist_size, double at_repr_intake_bin_range, int AAR_hist_size, double AAR_hist_range, int Reproduce_hist_size, double Reproduce_hist_range, int genome_size_histogram_size)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Splits into a histograms a series of data of all the cells which reproduced.  <a href="#a882e846f4004e31b88e801514c932484"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a82222fece3a09085304948f8c338729f">returnMeanEnvelopeToEnvRatio</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the mean size of the surface of the environment taken by genotypes envelopes.  <a href="#a82222fece3a09085304948f8c338729f"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a3af5683a5e2d7c0ff3e8a726f2c5eb32">returnSTDEnvelopeToEnvRatio</a> ()</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Returns the standard deviation of the surface of the environment covered by genotypes envelopes.  <a href="#a3af5683a5e2d7c0ff3e8a726f2c5eb32"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222">writeGeneralDataToFile</a> (int k, double env)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Saving general system's properties to the file.  <a href="#aa26876007e770024928bd8c2748d2222"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#aeea43b7824caa0001a8fe248c3392017">cumulativeMutationNumbers</a> (int current_time_stamp)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Calculating the speed of evolution.  <a href="#aeea43b7824caa0001a8fe248c3392017"></a><br/></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a4227871d6d47201edf4450fd35882c53">clearMutationCounters</a> (int current_iteration)</td></tr>
<tr><td class="mdescLeft">&#160;</td><td class="mdescRight">Data collecting method. Clears all three record of all the cells regarding the number of mutations.  <a href="#a4227871d6d47201edf4450fd35882c53"></a><br/></td></tr>
<tr><td colspan="2"><h2><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="classCell.html">Cell</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a52801d9d0de539460a718b0c50eb9e32">meanGenomeSize</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a628b26f0fc25e8d334ba3d9a96dc43eb">STD_GenomeSize</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">unsigned long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classEcosystem.html#a972b9932e42c702002c7ffc208e009b5">startOfMutationCounter</a></td></tr>
</table>
<hr/><a name="details" id="details"></a><h2>Detailed Description</h2>
<div class="textblock"><p>This is 'The <a class="el" href="classEcosystem.html" title="This is &#39;The Ecosystem&#39;. A place where things are happening.">Ecosystem</a>'. A place where things are happening. </p>
<p>Stores a vector (<a href="http://www.cplusplus.com/reference/stl/vector/">STL container</a>) of cell objects. Has methods to set up a population of cells and later feed, kill and reproduce them. Has a lot methods to harvest data. </p>

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00052">52</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>
</div><hr/><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="aad52d5b02f1e2f0ffd16c72825dee140"></a><!-- doxytag: member="Ecosystem::Ecosystem" ref="aad52d5b02f1e2f0ffd16c72825dee140" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classEcosystem.html#aad52d5b02f1e2f0ffd16c72825dee140">Ecosystem::Ecosystem</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Constructor - sets an ecosystem, but with no values. </p>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00040">40</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                     {
}
</pre></div>
</div>
</div>
<hr/><h2>Member Function Documentation</h2>
<a class="anchor" id="a4227871d6d47201edf4450fd35882c53"></a><!-- doxytag: member="Ecosystem::clearMutationCounters" ref="a4227871d6d47201edf4450fd35882c53" args="(int current_iteration)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a4227871d6d47201edf4450fd35882c53">Ecosystem::clearMutationCounters</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>current_iteration</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Clears all three record of all the cells regarding the number of mutations. </p>
<p>Iterates through the population setting all the mutation counters of all three mutation types to zero. Can be called at any time. </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">current_iteration</td><td>- number of the current iteration. </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l01036">1036</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                           {
    <span class="keywordtype">int</span> PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">if</span>(PopSize &gt; 0){
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; PopSize; i++) {
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalPoinMutationNumber = 0;
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalDuplicationNumber = 0;
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalDeletionNumber = 0;
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalHGT_number = 0;
        }
        <a class="code" href="classEcosystem.html#a972b9932e42c702002c7ffc208e009b5">startOfMutationCounter</a> = current_iteration;
    }
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a4227871d6d47201edf4450fd35882c53_icgraph.png" border="0" usemap="#classEcosystem_a4227871d6d47201edf4450fd35882c53_icgraph" alt=""/></div>
<map name="classEcosystem_a4227871d6d47201edf4450fd35882c53_icgraph" id="classEcosystem_a4227871d6d47201edf4450fd35882c53_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="289,5,340,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aeea43b7824caa0001a8fe248c3392017"></a><!-- doxytag: member="Ecosystem::cumulativeMutationNumbers" ref="aeea43b7824caa0001a8fe248c3392017" args="(int current_time_stamp)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#aeea43b7824caa0001a8fe248c3392017">Ecosystem::cumulativeMutationNumbers</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>current_time_stamp</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Calculating the speed of evolution. </p>
<p>Iterates through the population writing to a file cells' Own Tag and the total numbers of all three kinds of mutation (point mutation, gene duplication gene deletion) which lead to the current genetic lineages. Time stamps must be given by the programmer.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">current_time_stamp</td><td>- a time stamp representing the current iteration number </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l01004">1004</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                                {
    <span class="keywordtype">int</span> PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    std::ofstream MutationRecords;
    MutationRecords.open(<span class="stringliteral">&quot;MutationRecords.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">if</span>(PopSize &gt; 0){
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; PopSize; i++) {
            MutationRecords &lt;&lt; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].OwnTag &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; \
                    PopulationOfCells[i].CellsGenotype.TotalPoinMutationNumber &lt;&lt; \
                    <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalDuplicationNumber &lt;&lt; \
                    <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalDeletionNumber &lt;&lt; \
                    <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.TotalHGT_number &lt;&lt; \
                    std::endl;
        }

    } <span class="keywordflow">else</span> {
        MutationRecords &lt;&lt; <span class="stringliteral">&quot;-1 All is dead!&quot;</span> &lt;&lt; std::endl;
    }
    MutationRecords &lt;&lt; <span class="stringliteral">&quot;#start_of_the_mutation_counter: &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a972b9932e42c702002c7ffc208e009b5">startOfMutationCounter</a> &lt;&lt; \
        std::endl;
    MutationRecords &lt;&lt; <span class="stringliteral">&quot;#time_at_data_collecting: &quot;</span> &lt;&lt; current_time_stamp &lt;&lt; std::endl;
    MutationRecords.close();
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_aeea43b7824caa0001a8fe248c3392017_icgraph.png" border="0" usemap="#classEcosystem_aeea43b7824caa0001a8fe248c3392017_icgraph" alt=""/></div>
<map name="classEcosystem_aeea43b7824caa0001a8fe248c3392017_icgraph" id="classEcosystem_aeea43b7824caa0001a8fe248c3392017_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="327,5,377,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a144a0594e881ad7b36b64faccf1d9b4c"></a><!-- doxytag: member="Ecosystem::feedCells" ref="a144a0594e881ad7b36b64faccf1d9b4c" args="(std::string mode_of_feeding, double Env, double gene_replication_cost, int core_genes, double metabolic_costs, double gain_from_the_genotype_ratio)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a144a0594e881ad7b36b64faccf1d9b4c">Ecosystem::feedCells</a> </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>mode_of_feeding</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>Env</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>gene_replication_cost</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>core_genes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>metabolic_costs</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>gain_from_the_genotype_ratio</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Feeds cells. Has two modes of feeding: (p) proportionally to cell's demand for resource and to resource availability and (s) sequential feeding when cells are aligned in a random queue and fed one by one until depleting resources (not all cells are fed then). </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">mode_of_feeding</td><td>- a string indicating the mode of feeding </td></tr>
    <tr><td class="paramname">Env</td><td>- current value of the environmental conditions </td></tr>
    <tr><td class="paramname">gene_replication_cost</td><td>- cost of having one metabolic gene </td></tr>
    <tr><td class="paramname">core_genes</td><td>- number of non-reducible genes </td></tr>
    <tr><td class="paramname">metabolic_costs</td><td>- constant metabolic cost per cell </td></tr>
    <tr><td class="paramname">gain_from_the_genotype_ratio</td><td>- the maximal permitted uptake rate </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00104">104</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                             {
    <span class="comment">// number of cells in population</span>
    <span class="keywordtype">int</span> PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <a class="code" href="classrngEngine.html" title="A singleton class which runs the Marsenne Twister random number generator based on the C code of the ...">rngEngine</a>* pRandomNumberGenerator = <a class="code" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG.">rngEngine::getInstance</a>();
    <span class="keywordtype">double</span> SharedBetweenFactor = 0.0;

    <span class="keywordflow">if</span> (mode_of_feeding.compare(<span class="stringliteral">&quot;p&quot;</span>) == 0) {
        <span class="comment">// array of the demands of all the cells in this round</span>
        <span class="keywordtype">double</span> CellsDemandsInThisRound[PopSize];
        <span class="comment">// costs of living - cells lose resource and free resource increase</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; PopSize; k++) {
            <span class="comment">// evaluation of internal resources of a cell (just costs)</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].evaluateInternalResourcesCost(gene_replication_cost,
                    core_genes, metabolic_costs);
            <span class="comment">// ageing cells</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].ageingACell();
            <span class="comment">// evaluation of free resources</span>
            <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> -= <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].recentLoss;
            <span class="comment">// zeroing cell&#39;s recent gain of resources</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].zeroRecentGainOfCell();
            <span class="comment">// evaluation how much a cell can gain in this round of feeding</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].evaluateGrowthRateFromGenotypeOnePointEnv(Env);
            <span class="comment">// Demand of a cell for resources</span>
            CellsDemandsInThisRound[k] = gain_from_the_genotype_ratio \
                * <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].CellsGrossGrowthRateAtTheMoment;
            SharedBetweenFactor += CellsDemandsInThisRound[k];
        }
        SharedBetweenFactor = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> / SharedBetweenFactor;

        <span class="comment">// the feeding itself:</span>
        <span class="comment">// cells are being fed and Free Resources are being depleted</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; PopSize; j++) {
            <span class="comment">//evaluation of the gain of resources of a cell</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].evaluateInternalResourcesGainProportional
                    (gain_from_the_genotype_ratio, SharedBetweenFactor);
            <span class="comment">// evaluation of free resources</span>
            <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> -= <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].recentGain;
        }
    }
    <span class="keywordflow">if</span> (mode_of_feeding.compare(<span class="stringliteral">&quot;s&quot;</span>) == 0) {
        <span class="comment">// initialising a queuing array which will decide who will be fed first</span>
        <span class="keywordtype">int</span> ShuffleArr[PopSize];
        <span class="comment">// filling a queuing array which will decide who will be fed first</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; PopSize; i++) {
            ShuffleArr[i] = i;
        }
        <span class="comment">// Shuffling elements of queuing array by randomly exchanging each with one other</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; (PopSize - 1); i++) {
            <span class="comment">// Random remaining position</span>
            <span class="keywordtype">int</span> r = i + (((int) \
                (10000000.0 * pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>())) % (PopSize - i));
            <span class="keywordtype">int</span> temp = ShuffleArr[i];
            ShuffleArr[i] = ShuffleArr[r];
            ShuffleArr[r] = temp;
        }
        <span class="comment">// costs of living - cells lose resource and free resource increase</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; PopSize; k++) {
            <span class="comment">// evaluation of internal resources of a cell (just costs)</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].evaluateInternalResourcesCost(gene_replication_cost,
                    core_genes, metabolic_costs);
            <span class="comment">// evaluation of free resources</span>
            <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> -= <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].recentLoss;
            <span class="comment">// ageing cells</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].ageingACell();
            <span class="comment">// zeroing cell&#39;s recent gain of resources</span>
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].zeroRecentGainOfCell();
        }
        <span class="comment">// the feeding itself:</span>
        <span class="comment">// cells are being fed and Free Resources are being depleted</span>
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; PopSize; j++) {
            <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> &gt;= <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[ShuffleArr[j]].CellsResource) {
                <span class="comment">// evaluation how much a cell can gain in this round of feeding</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[ShuffleArr[j]].evaluateGrowthRateFromGenotypeOnePointEnv(Env);
                <span class="comment">//evaluation of the gain of resources of a cell</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[ShuffleArr[j]].evaluateInternalResourcesGainSequential
                        (gain_from_the_genotype_ratio);
                <span class="comment">// evaluation of free resources</span>
                <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> -= <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[ShuffleArr[j]].recentGain;
            } <span class="keywordflow">else</span> {
                <span class="keywordflow">break</span>;
            }
        }
    }
    <span class="keywordflow">if</span> ((mode_of_feeding.compare(<span class="stringliteral">&quot;p&quot;</span>) != 0) &amp;&amp; (mode_of_feeding.compare(<span class="stringliteral">&quot;s&quot;</span>) != 0)) {
        std::cout &lt;&lt; <span class="stringliteral">&quot;Mode can be only p (proportional) or s (sequential)&quot;</span>\
            &lt;&lt; std::endl;
    }
}
</pre></div>
<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_cgraph.png" border="0" usemap="#classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_cgraph" alt=""/></div>
<map name="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_cgraph" id="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_cgraph">
<area shape="rect" id="node3" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG." alt="" coords="225,5,388,35"/><area shape="rect" id="node7" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number." alt="" coords="209,58,404,89"/><area shape="rect" id="node5" href="classrngEngine.html#aba770bb62302fb289d272c996163135f" title="Core method. Constructor of the RNG." alt="" coords="453,5,608,35"/><area shape="rect" id="node9" href="classrngEngine.html#a1b269e4352e5eae019647eadd4088bac" title="Marsenne Twister random number generator&#39;s guts." alt="" coords="456,58,605,89"/><area shape="rect" id="node11" href="classrngEngine.html#a21840f29a1ea114ddd251553e339f168" title="Seeding the Marsenne Twister random number generator." alt="" coords="657,58,807,89"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_icgraph.png" border="0" usemap="#classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_icgraph" alt=""/></div>
<map name="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_icgraph" id="classEcosystem_a144a0594e881ad7b36b64faccf1d9b4c_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="209,5,260,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a611ff2fecef4a34dcb73657ad572dbf8"></a><!-- doxytag: member="Ecosystem::histCellsAge" ref="a611ff2fecef4a34dcb73657ad572dbf8" args="(int histSize)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a611ff2fecef4a34dcb73657ad572dbf8">Ecosystem::histCellsAge</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>HistSize</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns a histogram of ages of cells. </p>
<p>Splits into a histogram cells' age data. Writes that to file. File name is fixed in this method what obviously limits the number of model executions in one direData collecting method. ctory. </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">HistSize</td><td>- the number of bins in the histogram of the age structure </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>void </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00545">545</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                         {
    <span class="comment">// setting up histogram for collecting data of genome sizes</span>
    <span class="comment">//  int HistSize = (int) (MAX_GENOTYPE_INITIALIZED + 0.1 * MAX_GENOTYPE_INITIALIZED) - 1;</span>
    gsl_histogram * hist = gsl_histogram_alloc(HistSize);
    <span class="keywordtype">double</span> ranges[HistSize + 1];
    <span class="keywordtype">double</span> Range = 0.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= HistSize; j++) {
        ranges[j] = Range;
        Range += 10.0;
    }
    gsl_histogram_set_ranges(hist, ranges, HistSize + 1);
    <span class="comment">// felling in histogram with genotype size data</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist, (<span class="keywordtype">double</span>) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsAge);
    }
    <span class="comment">// writing genome data to a file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;CellsAgeData.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; HistSize; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; std::endl;
    histFile.close();
    gsl_histogram_free(hist);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a611ff2fecef4a34dcb73657ad572dbf8_icgraph.png" border="0" usemap="#classEcosystem_a611ff2fecef4a34dcb73657ad572dbf8_icgraph" alt=""/></div>
<map name="classEcosystem_a611ff2fecef4a34dcb73657ad572dbf8_icgraph" id="classEcosystem_a611ff2fecef4a34dcb73657ad572dbf8_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="228,5,279,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a29523e18d64fa67c8aebb9a99aff5957"></a><!-- doxytag: member="Ecosystem::histGenomeSize" ref="a29523e18d64fa67c8aebb9a99aff5957" args="(int genome_size_histogram_size)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a29523e18d64fa67c8aebb9a99aff5957">Ecosystem::histGenomeSize</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>genome_size_histogram_size</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Writes to file a histogram of sizes of genotypes in given time step. </p>
<p>File name is fixed in this method what obviously limits the number of model executions in one directory. </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">genome_size_histogram_size</td><td>- maximal size of genome you expect the simulation will produce (will be used to produce time series of histograms) </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00508">508</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                             {
    <span class="comment">// setting up histogram for collecting data of genome sizes</span>
    gsl_histogram * hist_genome = gsl_histogram_alloc(genome_size_histogram_size);
    <span class="keywordtype">double</span> ranges_genome[genome_size_histogram_size + 1];
    <span class="keywordtype">double</span> Range_genome = 1.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= genome_size_histogram_size; j++) {
        ranges_genome[j] = Range_genome;
        Range_genome += 1.0;
    }
    gsl_histogram_set_ranges(hist_genome, ranges_genome, genome_size_histogram_size + 1);
    <span class="comment">// felling in histogram with genotype size data</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist_genome,
                (<span class="keywordtype">double</span>) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.GenotypeValues.size());
    }
    <span class="comment">// writing genome data to a file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;GenomeSizeData.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; genome_size_histogram_size; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist_genome, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; <span class="stringliteral">&quot;\n&quot;</span>;
    histFile.close();
    gsl_histogram_free(hist_genome);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a29523e18d64fa67c8aebb9a99aff5957_icgraph.png" border="0" usemap="#classEcosystem_a29523e18d64fa67c8aebb9a99aff5957_icgraph" alt=""/></div>
<map name="classEcosystem_a29523e18d64fa67c8aebb9a99aff5957_icgraph" id="classEcosystem_a29523e18d64fa67c8aebb9a99aff5957_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="252,5,303,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a8ab8b96cd612eebec075a5a98feeec29"></a><!-- doxytag: member="Ecosystem::histGenotypeEnvelopeMaximum" ref="a8ab8b96cd612eebec075a5a98feeec29" args="(int maximum_of_genotype_hist_size, double maximum_of_genotype_hist_range)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a8ab8b96cd612eebec075a5a98feeec29">Ecosystem::histGenotypeEnvelopeMaximum</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>maximum_of_genotype_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>maximum_of_genotype_hist_range</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns a histogram of maximums of genotypes envelopes. </p>
<p>Splits into a histogram maximums of envelopes of genotypes. Writes that to file.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">maximum_of_genotype_hist_size</td><td>- number of bins in the genome size histogram </td></tr>
    <tr><td class="paramname">maximum_of_genotype_hist_range</td><td>- width of a bin in this histogram </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00710">710</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                               {
    <span class="comment">// setting up histogram for collecting data</span>
    gsl_histogram * hist = gsl_histogram_alloc(maximum_of_genotype_hist_size);
    <span class="keywordtype">double</span> ranges[maximum_of_genotype_hist_size + 1];
    <span class="keywordtype">double</span> Range = 0.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= maximum_of_genotype_hist_size; j++) {
        ranges[j] = Range;
        Range += maximum_of_genotype_hist_range;
    }
    gsl_histogram_set_ranges(hist, ranges, maximum_of_genotype_hist_size + 1);
    <span class="comment">// filling in histogram</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist, <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.MaxGrowthRate);
    }
    <span class="comment">// writing data to a histogram file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;FrameMaxData.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; maximum_of_genotype_hist_size; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; std::endl;
    histFile.close();
    gsl_histogram_free(hist);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a8ab8b96cd612eebec075a5a98feeec29_icgraph.png" border="0" usemap="#classEcosystem_a8ab8b96cd612eebec075a5a98feeec29_icgraph" alt=""/></div>
<map name="classEcosystem_a8ab8b96cd612eebec075a5a98feeec29_icgraph" id="classEcosystem_a8ab8b96cd612eebec075a5a98feeec29_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="348,5,399,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a5853f9762c4c91d40dfb05c959abd097"></a><!-- doxytag: member="Ecosystem::histGenotypeEnvelopeSurface" ref="a5853f9762c4c91d40dfb05c959abd097" args="(int envelope_of_genotype_hist_size, double envelope_of_genotype_hist_range)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a5853f9762c4c91d40dfb05c959abd097">Ecosystem::histGenotypeEnvelopeSurface</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>envelope_of_genotype_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>envelope_of_genotype_hist_range</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns a histogram of fractions of total environmental space taken by genotypes. </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">envelope_of_genotype_hist_size</td><td>- number of bins in the histogram of the fraction of the total env space taken by the genotype </td></tr>
    <tr><td class="paramname">envelope_of_genotype_hist_range</td><td>- width of a bin in this histogram </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00674">674</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                {
    <span class="comment">// setting up histogram for collecting data of frame sizes</span>
    gsl_histogram * hist = gsl_histogram_alloc(envelope_of_genotype_hist_size);
    <span class="keywordtype">double</span> ranges[envelope_of_genotype_hist_size + 1];
    <span class="keywordtype">double</span> Range = 0.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= envelope_of_genotype_hist_size; j++) {
        ranges[j] = Range;
        Range += envelope_of_genotype_hist_range;
    }
    gsl_histogram_set_ranges(hist, ranges, envelope_of_genotype_hist_size + 1);
    <span class="comment">// felling in histogram with genotype frame size to env surface ratio data</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist, <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.EnvSpaceFraction);
    }
    <span class="comment">// writing data to a histogram file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;FrameSizeData.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; envelope_of_genotype_hist_size; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; std::endl;
    histFile.close();
    gsl_histogram_free(hist);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a5853f9762c4c91d40dfb05c959abd097_icgraph.png" border="0" usemap="#classEcosystem_a5853f9762c4c91d40dfb05c959abd097_icgraph" alt=""/></div>
<map name="classEcosystem_a5853f9762c4c91d40dfb05c959abd097_icgraph" id="classEcosystem_a5853f9762c4c91d40dfb05c959abd097_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="337,5,388,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ab658ff43ae86399189569f9b00222c65"></a><!-- doxytag: member="Ecosystem::histResourceAllocatedInCells" ref="ab658ff43ae86399189569f9b00222c65" args="(int resource_alloc_hist_size, double resource_alloc_hist_range, double resource_alloc_hist_start)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#ab658ff43ae86399189569f9b00222c65">Ecosystem::histResourceAllocatedInCells</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>resource_alloc_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>resource_alloc_hist_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>resource_alloc_hist_start</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns a histogram of resources allocated in the cells. </p>
<p>Splits into a histogram resource allocated to all cells in the population. Writes that to file.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">resource_alloc_hist_size</td><td>- number of bins in the histogram </td></tr>
    <tr><td class="paramname">resource_alloc_hist_range</td><td>- width of a bin </td></tr>
    <tr><td class="paramname">resource_alloc_hist_start</td><td>- most-left value of the histogram </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00749">749</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                                            {
    <span class="comment">// setting up histogram for collecting data</span>
    gsl_histogram * hist = gsl_histogram_alloc(resource_alloc_hist_size);
    <span class="keywordtype">double</span> ranges[resource_alloc_hist_size + 1];
    <span class="keywordtype">double</span> Range = resource_alloc_hist_start;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= resource_alloc_hist_size; j++) {
        ranges[j] = Range;
        Range += resource_alloc_hist_range;
    }
    gsl_histogram_set_ranges(hist, ranges, resource_alloc_hist_size + 1);
    <span class="comment">// filling in histogram</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist, <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsResource);
    }
    <span class="comment">// writing data to a histogram file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;RecourceInCells.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; resource_alloc_hist_size; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; std::endl;
    histFile.close();
    gsl_histogram_free(hist);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_ab658ff43ae86399189569f9b00222c65_icgraph.png" border="0" usemap="#classEcosystem_ab658ff43ae86399189569f9b00222c65_icgraph" alt=""/></div>
<map name="classEcosystem_ab658ff43ae86399189569f9b00222c65_icgraph" id="classEcosystem_ab658ff43ae86399189569f9b00222c65_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="329,5,380,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="acd7c553337ea8a38ecb4a1572cf68b52"></a><!-- doxytag: member="Ecosystem::histResourceUptakenByCells" ref="acd7c553337ea8a38ecb4a1572cf68b52" args="(int resource_uptake_hist_size, double resource_uptake_hist_range)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#acd7c553337ea8a38ecb4a1572cf68b52">Ecosystem::histResourceUptakenByCells</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>resource_uptake_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>resource_uptake_hist_range</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns a histogram of resources allocated in the cells. </p>
<p>Splits into a histogram resource allocated to all cells in the population. Writes that to file.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">resource_uptake_hist_size</td><td>- number of bins in the histogram </td></tr>
    <tr><td class="paramname">resource_uptake_hist_range</td><td>- width of a bin </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00787">787</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                           {
    <span class="comment">// setting up histogram for collecting data</span>
    gsl_histogram * hist = gsl_histogram_alloc(resource_uptake_hist_size);
    <span class="keywordtype">double</span> ranges[resource_uptake_hist_size + 1];
    <span class="keywordtype">double</span> Range = 0.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= resource_uptake_hist_size; j++) {
        ranges[j] = Range;
        Range += resource_uptake_hist_range;
    }
    gsl_histogram_set_ranges(hist, ranges, resource_uptake_hist_size + 1);
    <span class="comment">// filling in histogram</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        gsl_histogram_increment(hist, <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].recentGain);
    }
    <span class="comment">// writing data to a histogram file</span>
    std::ofstream histFile;
    <span class="comment">// ooop, watch that line below when changing the file name!</span>
    histFile.open(<span class="stringliteral">&quot;RecourceUptakenCells.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; resource_uptake_hist_size; k++) {
        histFile &lt;&lt; gsl_histogram_get(hist, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile &lt;&lt; std::endl;
    histFile.close();
    gsl_histogram_free(hist);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_acd7c553337ea8a38ecb4a1572cf68b52_icgraph.png" border="0" usemap="#classEcosystem_acd7c553337ea8a38ecb4a1572cf68b52_icgraph" alt=""/></div>
<map name="classEcosystem_acd7c553337ea8a38ecb4a1572cf68b52_icgraph" id="classEcosystem_acd7c553337ea8a38ecb4a1572cf68b52_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="329,5,380,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a882e846f4004e31b88e801514c932484"></a><!-- doxytag: member="Ecosystem::histsAgeAndResourceAtReprod" ref="a882e846f4004e31b88e801514c932484" args="(int at_repr_intake_hist_size, double at_repr_intake_bin_range, int AAR_hist_size, double AAR_hist_range, int Reproduce_hist_size, double Reproduce_hist_range, int genome_size_histogram_size)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a882e846f4004e31b88e801514c932484">Ecosystem::histsAgeAndResourceAtReprod</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>at_repr_intake_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>at_repr_intake_bin_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>AAR_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>AAR_hist_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>Reproduce_hist_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>Reproduce_hist_range</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>genome_size_histogram_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Splits into a histograms a series of data of all the cells which reproduced. </p>
<p>Makes five histograms only for cells which reproduced: the average intake per unit of time of cells; the total gain per cell; cells' age at reproduction; number of reproductions of cells; gene number in cells' genotypes. Writes that to file.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">at_repr_intake_hist_size</td><td>- number of bins in the histogram of average intake per cell </td></tr>
    <tr><td class="paramname">at_repr_intake_bin_range</td><td>- width of bins in it </td></tr>
    <tr><td class="paramname">AAR_hist_size</td><td>- number of bins in age at reproduction histogram </td></tr>
    <tr><td class="paramname">AAR_hist_range</td><td>- width of a bin of age at reproduction histogram </td></tr>
    <tr><td class="paramname">Reproduce_hist_size</td><td>- number of bins in histogram of how many times a cell reproduced </td></tr>
    <tr><td class="paramname">Reproduce_hist_range</td><td>- width of a bin of histogram of how many times a cell reproduced </td></tr>
    <tr><td class="paramname">genome_size_histogram_size</td><td>- number of bins in genome size histogram </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00832">832</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                        {
    <span class="comment">// setting up histogram for collecting data</span>
    gsl_histogram * hist_intake = gsl_histogram_alloc(at_repr_intake_hist_size);
    gsl_histogram * hist_tot_gain = gsl_histogram_alloc(at_repr_intake_hist_size);
    gsl_histogram * hist_age_repr = gsl_histogram_alloc(AAR_hist_size);
    gsl_histogram * hist_num_repr = gsl_histogram_alloc(Reproduce_hist_size);
    gsl_histogram * hist_genome = gsl_histogram_alloc(genome_size_histogram_size);
    <span class="keywordtype">double</span> ranges[at_repr_intake_hist_size + 1];
    <span class="keywordtype">double</span> ranges_repr[AAR_hist_size + 1];
    <span class="keywordtype">double</span> ranges_num_repr[Reproduce_hist_size + 1];
    <span class="keywordtype">double</span> ranges_genome[genome_size_histogram_size + 1];
    <span class="keywordtype">double</span> Range = 0.0;
    <span class="keywordtype">double</span> Range_AAR = 0.0;
    <span class="keywordtype">double</span> Range_num_repr = 0.0;
    <span class="keywordtype">double</span> Range_genome = 1.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= at_repr_intake_hist_size; j++) {
        ranges[j] = Range;
        Range += at_repr_intake_bin_range;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= AAR_hist_size; j++) {
        ranges_repr[j] = Range_AAR;
        Range_AAR += AAR_hist_range;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= Reproduce_hist_size; j++) {
        ranges_num_repr[j] = Range_num_repr;
        Range_num_repr += Reproduce_hist_range;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt;= genome_size_histogram_size; j++) {
        ranges_genome[j] = Range_genome;
        Range_genome += 1.0;
    }
    gsl_histogram_set_ranges(hist_intake, ranges, at_repr_intake_hist_size + 1);
    gsl_histogram_set_ranges(hist_tot_gain, ranges, at_repr_intake_hist_size + 1);
    gsl_histogram_set_ranges(hist_age_repr, ranges_repr, AAR_hist_size + 1);
    gsl_histogram_set_ranges(hist_num_repr, ranges_num_repr, Reproduce_hist_size + 1);
    gsl_histogram_set_ranges(hist_genome, ranges_genome, genome_size_histogram_size + 1);
    <span class="comment">// filling in histogram</span>
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].DidReproduced == <span class="keyword">true</span>) {
            gsl_histogram_increment(hist_intake,
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].AvarageIntakeTillRepr);
            gsl_histogram_increment(hist_tot_gain,
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].AvarageTotalGainTillRepr);
            gsl_histogram_increment(hist_age_repr,
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].TimeSinceLastReprod);
            gsl_histogram_increment(hist_num_repr,
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].NumberOfReprod);
            gsl_histogram_increment(hist_genome,
                    (<span class="keywordtype">double</span>) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.GenotypeValues.size());
        }
    }
    <span class="comment">// writing data to a histogram file</span>
    std::ofstream histFile_intake;
    std::ofstream histFile_tot_gain;
    std::ofstream histFile_repr;
    std::ofstream histFile_num_repr;
    std::ofstream histFile_gene_num;
    <span class="comment">// ooop, watch that 4 line below when changing the file name!</span>
    histFile_intake.open(<span class="stringliteral">&quot;AvarIntakeAtRepr.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    histFile_tot_gain.open(<span class="stringliteral">&quot;AvarGainAtRepr.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    histFile_repr.open(<span class="stringliteral">&quot;AgeTillReproduction.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    histFile_num_repr.open(<span class="stringliteral">&quot;NumberOfReproductions.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    histFile_gene_num.open(<span class="stringliteral">&quot;GeneNumberOfRepr.dat&quot;</span>,
            std::ios::out | std::ios::ate | std::ios::app);
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; at_repr_intake_hist_size; k++) {
        histFile_intake &lt;&lt; gsl_histogram_get(hist_intake, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
        histFile_tot_gain &lt;&lt; gsl_histogram_get(hist_tot_gain, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> l = 0; l &lt; AAR_hist_size; l++) {
        histFile_repr &lt;&lt; gsl_histogram_get(hist_age_repr, l) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> m = 0; m &lt; Reproduce_hist_size; m++) {
        histFile_num_repr &lt;&lt; gsl_histogram_get(hist_num_repr, m) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> k = 0; k &lt; genome_size_histogram_size; k++) {
        histFile_gene_num &lt;&lt; gsl_histogram_get(hist_genome, k) &lt;&lt; <span class="stringliteral">&quot; &quot;</span>;
    }
    histFile_intake &lt;&lt; std::endl;
    histFile_tot_gain &lt;&lt; std::endl;
    histFile_repr &lt;&lt; std::endl;
    histFile_num_repr &lt;&lt; std::endl;
    histFile_gene_num &lt;&lt; std::endl;

    histFile_intake.close();
    histFile_tot_gain.close();
    histFile_repr.close();
    histFile_num_repr.close();
    histFile_gene_num.close();

    gsl_histogram_free(hist_intake);
    gsl_histogram_free(hist_tot_gain);
    gsl_histogram_free(hist_age_repr);
    gsl_histogram_free(hist_num_repr);
    gsl_histogram_free(hist_genome);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a882e846f4004e31b88e801514c932484_icgraph.png" border="0" usemap="#classEcosystem_a882e846f4004e31b88e801514c932484_icgraph" alt=""/></div>
<map name="classEcosystem_a882e846f4004e31b88e801514c932484_icgraph" id="classEcosystem_a882e846f4004e31b88e801514c932484_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="345,5,396,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a4dccae9a322022d09724501419df349f"></a><!-- doxytag: member="Ecosystem::horizontalGeneTransfer" ref="a4dccae9a322022d09724501419df349f" args="(double hgt_donor_prob, double hgt_gene_trans)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a4dccae9a322022d09724501419df349f">Ecosystem::horizontalGeneTransfer</a> </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hgt_donor_prob</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>hgt_gene_trans</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Does the horizontal gene transfer. </p>
<p>Called at each time step. Iterates through the population and for each cell in a population tosses if it's gonna be a gene donor. Then picks a cell which gonna be a gene acceptor and in the last step it tosses for each gene in the donor genotype if it will get transfered to the acceptors genotype. Uses <a href="http://www.cplusplus.com/reference/clibrary/cmath/floor/">cmath floor()</a> for number rounding.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">hgt_donor_prob</td><td>- probability of being a donor </td></tr>
    <tr><td class="paramname">hgt_gene_trans</td><td>- of one gene to get transfered </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00361">361</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                           {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">int</span> k;
    <span class="keywordtype">bool</span> did_hgt_happened;
    <span class="comment">// calling and initialising random number generator</span>
    <a class="code" href="classrngEngine.html" title="A singleton class which runs the Marsenne Twister random number generator based on the C code of the ...">rngEngine</a>* pRandomNumberGenerator = <a class="code" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG.">rngEngine::getInstance</a>();
    <a class="code" href="classTagging__system.html" title="A singleton class which generates an unique tag for a cell.">Tagging_system</a>* pTagging_system = <a class="code" href="classTagging__system.html#a80d575a3bbb747f9e1ff71877a937cba" title="Data collecting method. Gets instance of tagging system.">Tagging_system::getInstance</a>();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        <span class="comment">// tossing if a cell will be a donor</span>
        <span class="keywordflow">if</span> (pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>() &lt; hgt_donor_prob){
            <span class="comment">// picking an acceptor cell</span>
            <span class="keywordflow">do</span> {
                k = (int) floor(pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>() \
                    * LystSize);
            } <span class="keywordflow">while</span> ( k == i);
            <span class="keywordtype">int</span> GenoypeSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.GenotypeValues.size();
            did_hgt_happened = <span class="keyword">false</span>;
            <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; GenoypeSize; j++){
                <span class="comment">// tossing if a gene gets transfered </span>
                <span class="keywordflow">if</span> (pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>() &lt; hgt_gene_trans){
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].CellsGenotype.GenotypeValues.push_back
                    (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.GenotypeValues[j]);
                    <span class="comment">// incrementing the counter of accepted genes</span>
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].CellsGenotype.TotalHGT_number += 1;
                    did_hgt_happened = <span class="keyword">true</span>;
                }
                <span class="keywordflow">if</span> (did_hgt_happened == <span class="keyword">true</span>){
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[k].OwnTag = pTagging_system-&gt;<a class="code" href="classTagging__system.html#af57001cb7fe7d78d05076231552d1911" title="Data collecting method. Obtaining an unique tag.">getTag</a>();
                }
            }
        } 
    }  
}
</pre></div>
<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a4dccae9a322022d09724501419df349f_cgraph.png" border="0" usemap="#classEcosystem_a4dccae9a322022d09724501419df349f_cgraph" alt=""/></div>
<map name="classEcosystem_a4dccae9a322022d09724501419df349f_cgraph" id="classEcosystem_a4dccae9a322022d09724501419df349f_cgraph">
<area shape="rect" id="node3" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG." alt="" coords="317,5,480,35"/><area shape="rect" id="node7" href="classTagging__system.html#a80d575a3bbb747f9e1ff71877a937cba" title="Data collecting method. Gets instance of tagging system." alt="" coords="297,58,500,89"/><area shape="rect" id="node11" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number." alt="" coords="301,111,496,142"/><area shape="rect" id="node17" href="classTagging__system.html#af57001cb7fe7d78d05076231552d1911" title="Data collecting method. Obtaining an unique tag." alt="" coords="311,165,487,195"/><area shape="rect" id="node5" href="classrngEngine.html#aba770bb62302fb289d272c996163135f" title="Core method. Constructor of the RNG." alt="" coords="589,5,744,35"/><area shape="rect" id="node9" href="classTagging__system.html#a2d9f2ccd8121a0309e35151b5f9ea4da" title="Data collecting method. Constructor." alt="" coords="549,58,784,89"/><area shape="rect" id="node13" href="classrngEngine.html#a1b269e4352e5eae019647eadd4088bac" title="Marsenne Twister random number generator&#39;s guts." alt="" coords="592,111,741,142"/><area shape="rect" id="node15" href="classrngEngine.html#a21840f29a1ea114ddd251553e339f168" title="Seeding the Marsenne Twister random number generator." alt="" coords="833,111,983,142"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a4dccae9a322022d09724501419df349f_icgraph.png" border="0" usemap="#classEcosystem_a4dccae9a322022d09724501419df349f_icgraph" alt=""/></div>
<map name="classEcosystem_a4dccae9a322022d09724501419df349f_icgraph" id="classEcosystem_a4dccae9a322022d09724501419df349f_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="297,5,348,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a6549e81ea4daa38924a3791f649e1833"></a><!-- doxytag: member="Ecosystem::printNumberOfGenesInCells" ref="a6549e81ea4daa38924a3791f649e1833" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a6549e81ea4daa38924a3791f649e1833">Ecosystem::printNumberOfGenesInCells</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Prints a number of genes in genotypes of all Cells. </p>
<p>Test function. Not called in 'regular' runs. </p>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00401">401</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                          {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.printNumberOfGenes();
    }
}
</pre></div>
</div>
</div>
<a class="anchor" id="ae1f5162ddcbbec9933d5e2c5c62678fe"></a><!-- doxytag: member="Ecosystem::reproduceOrKillCell" ref="ae1f5162ddcbbec9933d5e2c5c62678fe" args="(std::string mode_of_feeding, double Env, double duplication_probability, double deletion_probability, double mutate_threshold, int minimum_genotype_size, double not_super_gene_param, double gene_space_width, double environment_resolution, double one_over_total_env_space_surf, double random_death_rate, double minimal_cells_resource, double reproduction_resourse_size, double disparity_at_reproduction)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#ae1f5162ddcbbec9933d5e2c5c62678fe">Ecosystem::reproduceOrKillCell</a> </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>mode_of_feeding</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>Env</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>duplication_probability</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>deletion_probability</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mutate_threshold</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>minimum_genotype_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>not_super_gene_param</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>gene_space_width</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>environment_resolution</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>one_over_total_env_space_surf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>random_death_rate</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>minimal_cells_resource</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>reproduction_resourse_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>disparity_at_reproduction</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Reproduces or/and kills cells. </p>
<p>Does three things: (a) kill cells 'by accident' (random death), (b) removes form population cells with not enough resources to live (killing weak ones), (c) reproduces those, who are big enough and mutates its offspring if necessary. Massively uses <a href="http://www.cplusplus.com/reference/stl/vector/">STL vector</a> methods.</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">mode_of_feeding</td><td>- a string indicating the mode of feeding </td></tr>
    <tr><td class="paramname">Env</td><td>- current value of the environmental conditions </td></tr>
    <tr><td class="paramname">duplication_probability</td><td>- gene duplication probability </td></tr>
    <tr><td class="paramname">deletion_probability</td><td>- gene deletion probability </td></tr>
    <tr><td class="paramname">mutate_threshold</td><td>- gene modification probability </td></tr>
    <tr><td class="paramname">minimum_genotype_size</td><td>- minimum permitted genotype size </td></tr>
    <tr><td class="paramname">not_super_gene_param</td><td>- surface under the Gaussian curve </td></tr>
    <tr><td class="paramname">gene_space_width</td><td>- the width of the env space genes can take </td></tr>
    <tr><td class="paramname">environment_resolution</td><td>- resolution of the env space, 1/total env </td></tr>
    <tr><td class="paramname">one_over_total_env_space_surf</td><td>- surface (const supporting computation) </td></tr>
    <tr><td class="paramname">random_death_rate</td><td>- probability a cell will randomly dead in this iteration </td></tr>
    <tr><td class="paramname">minimal_cells_resource</td><td>- resource below which cell dies </td></tr>
    <tr><td class="paramname">reproduction_resourse_size</td><td>- size at which the cell divides </td></tr>
    <tr><td class="paramname">disparity_at_reproduction</td><td>- if in proportional feeding mode this value says how uneven the resource division is during the cell division </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00220">220</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                          {
    <span class="comment">// calling and initialising random number generator</span>
    <a class="code" href="classrngEngine.html" title="A singleton class which runs the Marsenne Twister random number generator based on the C code of the ...">rngEngine</a>* pRandomNumberGenerator = <a class="code" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG.">rngEngine::getInstance</a>();
    <a class="code" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a> = 0;
    <a class="code" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a> = 0;
    <span class="keywordtype">int</span> PopSize;
    <span class="comment">// kill cells &#39;by accident&#39; (random death)</span>
    PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = PopSize - 1; i &gt;= 0; i--) {
        <span class="keywordflow">if</span> ((pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>()) &lt; random_death_rate) {
            <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> + <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsResource;
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.erase(<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.begin() + i);
            <span class="comment">// count number of killed</span>
            <a class="code" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a> += 1;
        }
    }
    <span class="comment">// kill the weak ones</span>
    PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = PopSize - 1; i &gt;= 0; i--) {
        <span class="keywordflow">if</span> ((<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsResource) &lt; minimal_cells_resource) {
            <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> + <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsResource;
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.erase(<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.begin() + i);
            <span class="comment">// count number of killed</span>
            <a class="code" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a> += 1;
        }
    }
    <span class="comment">// let the fittest win! (reproduction and mutation)</span>
    PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="comment">// sequential feeding:</span>
    <span class="keywordflow">if</span> (mode_of_feeding.compare(<span class="stringliteral">&quot;s&quot;</span>) == 0) {
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; PopSize; j++) {
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellReprodFlagFalse();
            <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource &gt; reproduction_resourse_size) {
                <span class="comment">// calculating the intake ratio for the reproducing cell</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAgeSinceLastReprod();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAvarageIntakeTillRepr();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAvarageTotalGainTillRepr();
                <span class="comment">// Dividing the resource between the mother and the offspring</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource = \
                        (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource) * 0.5;
                <span class="comment">// note the fact this cell has reproduced</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellReprodFlagTrue();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].incrementNumberOfReprod();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellsResAfterReprod();
                <span class="comment">// create a &#39;newborn&#39; cell</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.push_back(<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j]);
                <span class="comment">// set new cell&#39;s params to zero (ones which need to be zeroed)</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().resetParamsOfNewCell();
                <span class="comment">// inheriting mothers tag</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().MothersTag = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].OwnTag;
                <span class="comment">// mutate (if necessary) both cells</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].mutateCell(Env, duplication_probability,
                        deletion_probability, mutate_threshold, minimum_genotype_size,
                        not_super_gene_param, gene_space_width, environment_resolution,
                        one_over_total_env_space_surf);
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().mutateCell(Env, duplication_probability,
                        deletion_probability, mutate_threshold, minimum_genotype_size,
                        not_super_gene_param, gene_space_width, environment_resolution,
                        one_over_total_env_space_surf);
                <span class="comment">// count number of born to record cell&#39;s age at first reproduction</span>
                <a class="code" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a> += 1;
            }
        }
    }
    <span class="comment">// proportional feeding:</span>
    <span class="keywordflow">if</span> (mode_of_feeding.compare(<span class="stringliteral">&quot;p&quot;</span>) == 0) {
        <span class="keywordtype">double</span> Division_disparity;
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; PopSize; j++) {
            <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellReprodFlagFalse();
            <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource &gt; reproduction_resourse_size) {
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAgeSinceLastReprod();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAvarageIntakeTillRepr();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setAvarageTotalGainTillRepr();
                Division_disparity = 0.5 + disparity_at_reproduction \
                    * pRandomNumberGenerator-&gt;<a class="code" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number.">rngMTgetDauble</a>();
                <span class="comment">// note the fact this cell has reproduced</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellReprodFlagTrue();
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].incrementNumberOfReprod();
                <span class="comment">// create a &#39;newborn&#39; cell</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.push_back(<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j]);
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().CellsResource = (1.0 - Division_disparity) \
                    * (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource);
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource = Division_disparity \
                    * (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].CellsResource);
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].setCellsResAfterReprod();
                <span class="comment">// set new cell&#39;s params to zero</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().resetParamsOfNewCell();
                <span class="comment">// inheriting mothers tag</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().MothersTag = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].OwnTag;
                <span class="comment">// mutate (if necessary) both cells</span>
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].mutateCell(Env, duplication_probability,
                        deletion_probability, mutate_threshold, minimum_genotype_size,
                        not_super_gene_param, gene_space_width, environment_resolution,
                        one_over_total_env_space_surf);
                <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().mutateCell(Env, duplication_probability,
                        deletion_probability, mutate_threshold, minimum_genotype_size,
                        not_super_gene_param, gene_space_width, environment_resolution,
                        one_over_total_env_space_surf);
                <span class="comment">// count number of born</span>
                <a class="code" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a> += 1;
            }
        }
    }
}
</pre></div>
<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_cgraph.png" border="0" usemap="#classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_cgraph" alt=""/></div>
<map name="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_cgraph" id="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_cgraph">
<area shape="rect" id="node3" href="classrngEngine.html#aee64f40e07926185f95052cdc7bf18e3" title="Core method. Gets instance of the RNG." alt="" coords="287,5,449,35"/><area shape="rect" id="node7" href="classrngEngine.html#a7a954bb07d2a3e8f20b04f14d5f1a59b" title="Generates a random double precision number." alt="" coords="271,58,465,89"/><area shape="rect" id="node5" href="classrngEngine.html#aba770bb62302fb289d272c996163135f" title="Core method. Constructor of the RNG." alt="" coords="515,5,669,35"/><area shape="rect" id="node9" href="classrngEngine.html#a1b269e4352e5eae019647eadd4088bac" title="Marsenne Twister random number generator&#39;s guts." alt="" coords="517,58,667,89"/><area shape="rect" id="node11" href="classrngEngine.html#a21840f29a1ea114ddd251553e339f168" title="Seeding the Marsenne Twister random number generator." alt="" coords="719,58,868,89"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_icgraph.png" border="0" usemap="#classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_icgraph" alt=""/></div>
<map name="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_icgraph" id="classEcosystem_ae1f5162ddcbbec9933d5e2c5c62678fe_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="271,5,321,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a29837c080dcab24bb549b54f726b7836"></a><!-- doxytag: member="Ecosystem::returnGenesShannonIndex" ref="a29837c080dcab24bb549b54f726b7836" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a29837c080dcab24bb549b54f726b7836">Ecosystem::returnGenesShannonIndex</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the Shannon Index for genes in the population. Each gene, regardless in which cell is located, is treated as an independent entity. Uses <a href="http://www.gnu.org/software/gsl/manual/html_node/Logarithm-and-Related-Functions.html">GNU Scientific Library logarithm function</a> </p>
<dl class="return"><dt><b>Returns:</b></dt><dd>Shannon Index value for diversity of genes in population. </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00624">624</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                          {
    <span class="keywordtype">int</span> LystSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">int</span> GenesInCell = 0;
    std::vector&lt;Gene&gt; AllTheGenesLyst;
    <span class="comment">//creating a mega-genotype containing ALL the genes in a population</span>
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> q = 0; q &lt; LystSize; q++) {
        GenesInCell = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[q].CellsGenotype.GenotypeValues.size();
        <span class="keywordflow">for</span> (<span class="keywordtype">int</span> p = 0; p &lt; GenesInCell; p++) {
            AllTheGenesLyst.push_back(<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[q].CellsGenotype.
                GenotypeValues[p]);
        }
    }
    <span class="keywordtype">int</span> GeneCounter = AllTheGenesLyst.size();
    <span class="keywordtype">double</span> NumberOfGenes = (double) GeneCounter;
    <span class="keywordtype">bool</span> IfCountedLyst[GeneCounter];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> w = 0; w &lt; GeneCounter; w++) {
        IfCountedLyst[w] = <span class="keyword">false</span>;
    }
    
    <span class="keywordtype">double</span> abundance = 1.0;
    <span class="keywordtype">double</span> Summ = 0.0;
    
    <span class="comment">// calculating the Shannon Index for genes</span>
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; GeneCounter; i++) {
        <span class="keywordflow">if</span> (IfCountedLyst[i] == <span class="keyword">false</span>) {
            abundance = 1.0;
            <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = i + 1; j &lt; GeneCounter - 1; j++) {
                <span class="keywordflow">if</span> (i != j &amp;&amp; IfCountedLyst[j] == <span class="keyword">false</span> \
                    &amp;&amp; AllTheGenesLyst[i].GeneValue[0] == AllTheGenesLyst[j].GeneValue[0] \
                    &amp;&amp; AllTheGenesLyst[i].GeneValue[2] == AllTheGenesLyst[j].GeneValue[2] \
                    &amp;&amp; AllTheGenesLyst[i].GeneValue[1] == AllTheGenesLyst[j].GeneValue[1] ){
                    abundance += 1.0;
                    IfCountedLyst[j] = <span class="keyword">true</span>;
                }
            }
            <span class="comment">// the proper calculations</span>
            Summ = Summ - ((abundance / NumberOfGenes) * gsl_sf_log(abundance / NumberOfGenes));
        }
    }
    <span class="keywordflow">return</span> Summ;
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a29837c080dcab24bb549b54f726b7836_icgraph.png" border="0" usemap="#classEcosystem_a29837c080dcab24bb549b54f726b7836_icgraph" alt=""/></div>
<map name="classEcosystem_a29837c080dcab24bb549b54f726b7836_icgraph" id="classEcosystem_a29837c080dcab24bb549b54f726b7836_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="317,5,560,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="609,5,660,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a6e753a4d44bae33695219d15a8324930"></a><!-- doxytag: member="Ecosystem::returnMeanAgeOfCells" ref="a6e753a4d44bae33695219d15a8324930" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a6e753a4d44bae33695219d15a8324930">Ecosystem::returnMeanAgeOfCells</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the mean age of all cells in ecosystem. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- mean age of cells (double) </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00456">456</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                       {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> AgeOfAllCells[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        AgeOfAllCells[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsAge;
    }
    <span class="keywordflow">return</span> gsl_stats_mean(AgeOfAllCells, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a6e753a4d44bae33695219d15a8324930_icgraph.png" border="0" usemap="#classEcosystem_a6e753a4d44bae33695219d15a8324930_icgraph" alt=""/></div>
<map name="classEcosystem_a6e753a4d44bae33695219d15a8324930_icgraph" id="classEcosystem_a6e753a4d44bae33695219d15a8324930_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="291,5,533,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="583,5,633,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a82222fece3a09085304948f8c338729f"></a><!-- doxytag: member="Ecosystem::returnMeanEnvelopeToEnvRatio" ref="a82222fece3a09085304948f8c338729f" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a82222fece3a09085304948f8c338729f">Ecosystem::returnMeanEnvelopeToEnvRatio</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the mean size of the surface of the environment taken by genotypes envelopes. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- mean envelope to environment ratio </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00947">947</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                               {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> EnvelopeToEnvRatio[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        EnvelopeToEnvRatio[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.EnvSpaceFraction;
    }
    <span class="keywordflow">return</span> gsl_stats_mean(EnvelopeToEnvRatio, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a82222fece3a09085304948f8c338729f_icgraph.png" border="0" usemap="#classEcosystem_a82222fece3a09085304948f8c338729f_icgraph" alt=""/></div>
<map name="classEcosystem_a82222fece3a09085304948f8c338729f_icgraph" id="classEcosystem_a82222fece3a09085304948f8c338729f_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="347,5,589,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="639,5,689,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ae4373e033efacb2fc1d8be6b5521880c"></a><!-- doxytag: member="Ecosystem::returnMeanNumberOfGenesInCells" ref="ae4373e033efacb2fc1d8be6b5521880c" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#ae4373e033efacb2fc1d8be6b5521880c">Ecosystem::returnMeanNumberOfGenesInCells</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the mean number of genes in genotypes of all the cells. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- mean number of genes (double) </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00418">418</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                 {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> GenotypeLength[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        GenotypeLength[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.returnNumberOfGenes();
    }
    <span class="keywordflow">return</span> gsl_stats_mean(GenotypeLength, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_ae4373e033efacb2fc1d8be6b5521880c_icgraph.png" border="0" usemap="#classEcosystem_ae4373e033efacb2fc1d8be6b5521880c_icgraph" alt=""/></div>
<map name="classEcosystem_ae4373e033efacb2fc1d8be6b5521880c_icgraph" id="classEcosystem_ae4373e033efacb2fc1d8be6b5521880c_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="368,5,611,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="660,5,711,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="aff0fcbd5b5fbfa21fd40401d35a5a35b"></a><!-- doxytag: member="Ecosystem::returnPopulationShannonIndex" ref="aff0fcbd5b5fbfa21fd40401d35a5a35b" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#aff0fcbd5b5fbfa21fd40401d35a5a35b">Ecosystem::returnPopulationShannonIndex</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the Shannon index value for population diversity. Cells which have even one gene different will be accounted as members of different clonal strains. </p>
<p>Calculates Shannon index in a given time. Uses <a href="http://www.gnu.org/software/gsl/manual/html_node/Logarithm-and-Related-Functions.html">GNU Scientific Library logarithm function</a></p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- Shannon index value (double) for population biodiversity. </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00584">584</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                               {
    <span class="keywordtype">int</span> LystSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> popSize = (double) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> abundance = 1.0;
    <span class="keywordtype">double</span> Summ = 0.0;
<span class="comment">//    double speciesRichness = 0.0;</span>
    <span class="comment">// zeroing &quot;if already considered&quot; counter</span>
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].TestIfCounted = <span class="keyword">false</span>;
    }
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].TestIfCounted == <span class="keyword">false</span>) {
            abundance = 1.0;
            <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = i + 1; j &lt; LystSize - 1; j++) {
                <span class="keywordflow">if</span> (i != j &amp;&amp; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].TestIfCounted == <span class="keyword">false</span> \
                        &amp;&amp; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].OwnTag == <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].OwnTag) {
                    abundance += 1.0;
                    <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[j].TestIfCounted = <span class="keyword">true</span>;
                }
            }
            <span class="comment">// the proper calculations</span>
            Summ = Summ - ((abundance / popSize) * gsl_sf_log(abundance / popSize));
<span class="comment">//            speciesRichness += 1.0;</span>
        }
    }
    <span class="keywordflow">return</span> Summ;
    <span class="comment">// Term with correction seems not to be the case, as this population is a complete</span>
    <span class="comment">// set of ALL individuals, not a sample from a larger population.</span>
    <span class="comment">//    return Summ - ((speciesRichness - 1.0) / (2.0 * popSize));</span>
}
</pre></div>
</div>
</div>
<a class="anchor" id="a5817a0df8a9753d2fb21fe2a465b1364"></a><!-- doxytag: member="Ecosystem::returnsSummOfCellResources" ref="a5817a0df8a9753d2fb21fe2a465b1364" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a5817a0df8a9753d2fb21fe2a465b1364">Ecosystem::returnsSummOfCellResources</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns returns the sum of all resources allocated in all the cells. </p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- the sum of resources in the cells (double) </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00490">490</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                             {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> Summ = 0.0;
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        Summ = Summ + <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsResource;
    }
    <span class="keywordflow">return</span> Summ;
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a5817a0df8a9753d2fb21fe2a465b1364_icgraph.png" border="0" usemap="#classEcosystem_a5817a0df8a9753d2fb21fe2a465b1364_icgraph" alt=""/></div>
<map name="classEcosystem_a5817a0df8a9753d2fb21fe2a465b1364_icgraph" id="classEcosystem_a5817a0df8a9753d2fb21fe2a465b1364_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="336,5,579,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="628,5,679,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a59bedd0e20191831596d1f5a8f8f6ad2"></a><!-- doxytag: member="Ecosystem::returnSTDAgeOfCells" ref="a59bedd0e20191831596d1f5a8f8f6ad2" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a59bedd0e20191831596d1f5a8f8f6ad2">Ecosystem::returnSTDAgeOfCells</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the standard deviation of the age of cells. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- STD of the age of cells (double) </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00475">475</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                      {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> AgeOfAllCells[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        AgeOfAllCells[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsAge;
    }
    <span class="keywordflow">return</span> gsl_stats_sd(AgeOfAllCells, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a59bedd0e20191831596d1f5a8f8f6ad2_icgraph.png" border="0" usemap="#classEcosystem_a59bedd0e20191831596d1f5a8f8f6ad2_icgraph" alt=""/></div>
<map name="classEcosystem_a59bedd0e20191831596d1f5a8f8f6ad2_icgraph" id="classEcosystem_a59bedd0e20191831596d1f5a8f8f6ad2_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="283,5,525,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="575,5,625,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a3af5683a5e2d7c0ff3e8a726f2c5eb32"></a><!-- doxytag: member="Ecosystem::returnSTDEnvelopeToEnvRatio" ref="a3af5683a5e2d7c0ff3e8a726f2c5eb32" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a3af5683a5e2d7c0ff3e8a726f2c5eb32">Ecosystem::returnSTDEnvelopeToEnvRatio</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the standard deviation of the surface of the environment covered by genotypes envelopes. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- STD of envelope to environment ratio </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00966">966</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                              {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> EnvelopeToEnvRatio[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        EnvelopeToEnvRatio[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.EnvSpaceFraction;
    }
    <span class="keywordflow">return</span> gsl_stats_sd(EnvelopeToEnvRatio, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a3af5683a5e2d7c0ff3e8a726f2c5eb32_icgraph.png" border="0" usemap="#classEcosystem_a3af5683a5e2d7c0ff3e8a726f2c5eb32_icgraph" alt=""/></div>
<map name="classEcosystem_a3af5683a5e2d7c0ff3e8a726f2c5eb32_icgraph" id="classEcosystem_a3af5683a5e2d7c0ff3e8a726f2c5eb32_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="341,5,584,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="633,5,684,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a30fc9729432ba63f068f19effa42d8dd"></a><!-- doxytag: member="Ecosystem::returnStdOfGenesInCells" ref="a30fc9729432ba63f068f19effa42d8dd" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a30fc9729432ba63f068f19effa42d8dd">Ecosystem::returnStdOfGenesInCells</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Returns the standard deviation of the number of genes in genotypes of all the cells. </p>
<p>Uses procedures from <a href="http://www.gnu.org/software/gsl/manual/html_node/Mean-and-standard-deviation-and-variance.html">GNU Scientific Library</a> to calculate those means.</p>
<dl class="return"><dt><b>Returns:</b></dt><dd>- STD of the number of genes (double) </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00437">437</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                          {
    <span class="keywordtype">int</span> LystSize = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordtype">double</span> GenotypeLength[LystSize];
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; LystSize; i++) {
        GenotypeLength[i] = <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].CellsGenotype.returnNumberOfGenes();
    }
    <span class="keywordflow">return</span> gsl_stats_sd(GenotypeLength, 1, LystSize);
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a30fc9729432ba63f068f19effa42d8dd_icgraph.png" border="0" usemap="#classEcosystem_a30fc9729432ba63f068f19effa42d8dd_icgraph" alt=""/></div>
<map name="classEcosystem_a30fc9729432ba63f068f19effa42d8dd_icgraph" id="classEcosystem_a30fc9729432ba63f068f19effa42d8dd_icgraph">
<area shape="rect" id="node3" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222" title="Data collecting method. Saving general system&#39;s properties to the file." alt="" coords="304,5,547,35"/><area shape="rect" id="node5" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="596,5,647,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="a9c459456082bdcebebeaba67b276cfd0"></a><!-- doxytag: member="Ecosystem::setPopulation" ref="a9c459456082bdcebebeaba67b276cfd0" args="(double total_resource_pool, double Env, double minimal_cells_resource, double maximal_resourse_bonus_at_start, int min_genotype_initialized, int max_genotype_initialized, double not_super_gene_param, double gene_space_width, double environment_resolution, double one_over_total_env_space_surf)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#a9c459456082bdcebebeaba67b276cfd0">Ecosystem::setPopulation</a> </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>total_resource_pool</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>Env</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>minimal_cells_resource</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>maximal_resourse_bonus_at_start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>min_genotype_initialized</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>max_genotype_initialized</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>not_super_gene_param</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>gene_space_width</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>environment_resolution</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>one_over_total_env_space_surf</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Sets a new population in given Env conditions and of given size. </p>
<p>Generates cells and puts them in a vector (<a href="http://www.cplusplus.com/reference/stl/vector/">STL vector</a>) until reaches the limit of ecosystem capacity defined in defines.h file. Should be called ones in the run (in the very beginning).</p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">total_resource_pool</td><td>- total amount of resource in the whole environment </td></tr>
    <tr><td class="paramname">Env</td><td>- current value of he environmental conditions </td></tr>
    <tr><td class="paramname">minimal_cells_resource</td><td>- minimal resource below which cell dies </td></tr>
    <tr><td class="paramname">maximal_resourse_bonus_at_start</td><td>- maximal bonus a cell cat get at start to be above the minimal limit </td></tr>
    <tr><td class="paramname">min_genotype_initialized</td><td>- minimal genotype size at initialisation </td></tr>
    <tr><td class="paramname">max_genotype_initialized</td><td>- maximal genotype at initialisation </td></tr>
    <tr><td class="paramname">not_super_gene_param</td><td>- surface under the Gaussian curve </td></tr>
    <tr><td class="paramname">gene_space_width</td><td>- the width of the env space genes can take </td></tr>
    <tr><td class="paramname">environment_resolution</td><td>- resolution of the env space </td></tr>
    <tr><td class="paramname">one_over_total_env_space_surf</td><td>- 1/total env surface (const supporting computation </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00063">63</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                                             {
    <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> = total_resource_pool;
    <a class="code" href="classEcosystem.html#a972b9932e42c702002c7ffc208e009b5">startOfMutationCounter</a> = 0;
    <span class="keywordtype">double</span> tempEnv = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a>;
    <a class="code" href="classCell.html" title="Represents one cell. This is the agent of the model.">Cell</a> tempCell;
    <span class="comment">// allocating a population of cells depending on the capacity of Environment</span>
    <span class="keywordflow">while</span> (<a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> &gt; 0.0) {
        tempEnv = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a>;
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.push_back(tempCell);
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().setBrandNewCell(Env, minimal_cells_resource,
                maximal_resourse_bonus_at_start, min_genotype_initialized,
                max_genotype_initialized, not_super_gene_param, gene_space_width,
                environment_resolution, one_over_total_env_space_surf);
        <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> = <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> - <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.back().CellsResource;
    }
    <span class="comment">// preventing the Free Resources from being a negative number</span>
    <span class="keywordflow">if</span> (<a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> &lt; 0.0) {
        <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> = tempEnv;
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.pop_back();
    }
    <a class="code" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a> = 0;
    <a class="code" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a> = 0;
}
</pre></div>
<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_a9c459456082bdcebebeaba67b276cfd0_icgraph.png" border="0" usemap="#classEcosystem_a9c459456082bdcebebeaba67b276cfd0_icgraph" alt=""/></div>
<map name="classEcosystem_a9c459456082bdcebebeaba67b276cfd0_icgraph" id="classEcosystem_a9c459456082bdcebebeaba67b276cfd0_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="233,5,284,35"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ace4318bc4766978edea2f73ebcaf98f8"></a><!-- doxytag: member="Ecosystem::updateGrowthRatesOfCellsInPopulation" ref="ace4318bc4766978edea2f73ebcaf98f8" args="(double Env)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#ace4318bc4766978edea2f73ebcaf98f8">Ecosystem::updateGrowthRatesOfCellsInPopulation</a> </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>Env</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Core method. Updates growth rates of all the cells in the ecosystem. </p>
<p>Called when environmental conditions change. Updates growth rates of all the cells in the ecosystem basing on theirs genotypes properties. For the sake of model's computational complexity should not be called in regular run (there are better way to update that before feeding). </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">Env</td><td>- current value of the environmental conditions. </td></tr>
  </table>
  </dd>
</dl>
<dl class="return"><dt><b>Returns:</b></dt><dd>void </dd></dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00341">341</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                               {
    <span class="keywordtype">int</span> PopSize = (int) <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size();
    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; PopSize; i++) {
        <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>[i].evaluateGrowthRateFromGenotypeOnePointEnv(Env);
    }
}
</pre></div>
</div>
</div>
<a class="anchor" id="aa26876007e770024928bd8c2748d2222"></a><!-- doxytag: member="Ecosystem::writeGeneralDataToFile" ref="aa26876007e770024928bd8c2748d2222" args="(int k, double env)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classEcosystem.html#aa26876007e770024928bd8c2748d2222">Ecosystem::writeGeneralDataToFile</a> </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>k</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>env</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Data collecting method. Saving general system's properties to the file. </p>
<dl class="params"><dt><b>Parameters:</b></dt><dd>
  <table class="params">
    <tr><td class="paramname">k</td><td>- number of iteration (time stamp) </td></tr>
    <tr><td class="paramname">env</td><td>- current value of the environmental conditions </td></tr>
  </table>
  </dd>
</dl>

<p>Definition at line <a class="el" href="ecosystem_8cpp_source.html#l00981">981</a> of file <a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a>.</p>
<div class="fragment"><pre class="fragment">                                                        {
    std::ofstream GeneralDataFile;
    GeneralDataFile.open(<span class="stringliteral">&quot;GeneralData.dat&quot;</span>, std::ios::out | std::ios::ate | std::ios::app);
    GeneralDataFile &lt;&lt; k &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; env &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#ae4373e033efacb2fc1d8be6b5521880c" title="Data collecting method. Returns the mean number of genes in genotypes of all the cells.">returnMeanNumberOfGenesInCells</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#a30fc9729432ba63f068f19effa42d8dd" title="Data collecting method. Returns the standard deviation of the number of genes in genotypes of all the...">returnStdOfGenesInCells</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">PopulationOfCells</a>.size() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#a5817a0df8a9753d2fb21fe2a465b1364" title="Data collecting method. Returns returns the sum of all resources allocated in all the cells...">returnsSummOfCellResources</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">EnvFreeResources</a> &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#a6e753a4d44bae33695219d15a8324930" title="Data collecting method. Returns the mean age of all cells in ecosystem.">returnMeanAgeOfCells</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a59bedd0e20191831596d1f5a8f8f6ad2" title="Data collecting method. Returns the standard deviation of the age of cells.">returnSTDAgeOfCells</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#a29837c080dcab24bb549b54f726b7836" title="Data collecting method. Returns the Shannon Index for genes in the population. Each gene...">returnGenesShannonIndex</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">numberOfBirths</a> &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">numberOfDeaths</a> &lt;&lt; <span class="stringliteral">&quot; &quot;</span> &lt;&lt; <a class="code" href="classEcosystem.html#a82222fece3a09085304948f8c338729f" title="Data collecting method. Returns the mean size of the surface of the environment taken by genotypes en...">returnMeanEnvelopeToEnvRatio</a>() &lt;&lt; <span class="stringliteral">&quot; &quot;</span> \
        &lt;&lt; <a class="code" href="classEcosystem.html#a3af5683a5e2d7c0ff3e8a726f2c5eb32" title="Data collecting method. Returns the standard deviation of the surface of the environment covered by g...">returnSTDEnvelopeToEnvRatio</a>() &lt;&lt; std::endl;
    GeneralDataFile.close();
}
</pre></div>
<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_aa26876007e770024928bd8c2748d2222_cgraph.png" border="0" usemap="#classEcosystem_aa26876007e770024928bd8c2748d2222_cgraph" alt=""/></div>
<map name="classEcosystem_aa26876007e770024928bd8c2748d2222_cgraph" id="classEcosystem_aa26876007e770024928bd8c2748d2222_cgraph">
<area shape="rect" id="node3" href="classEcosystem.html#ae4373e033efacb2fc1d8be6b5521880c" title="Data collecting method. Returns the mean number of genes in genotypes of all the cells." alt="" coords="296,5,611,35"/><area shape="rect" id="node5" href="classEcosystem.html#a30fc9729432ba63f068f19effa42d8dd" title="Data collecting method. Returns the standard deviation of the number of genes in genotypes of all the..." alt="" coords="328,58,579,89"/><area shape="rect" id="node7" href="classEcosystem.html#a5817a0df8a9753d2fb21fe2a465b1364" title="Data collecting method. Returns returns the sum of all resources allocated in all the cells..." alt="" coords="313,111,593,142"/><area shape="rect" id="node9" href="classEcosystem.html#a6e753a4d44bae33695219d15a8324930" title="Data collecting method. Returns the mean age of all cells in ecosystem." alt="" coords="336,165,571,195"/><area shape="rect" id="node11" href="classEcosystem.html#a59bedd0e20191831596d1f5a8f8f6ad2" title="Data collecting method. Returns the standard deviation of the age of cells." alt="" coords="340,218,567,249"/><area shape="rect" id="node13" href="classEcosystem.html#a29837c080dcab24bb549b54f726b7836" title="Data collecting method. Returns the Shannon Index for genes in the population. Each gene..." alt="" coords="321,271,585,302"/><area shape="rect" id="node15" href="classEcosystem.html#a82222fece3a09085304948f8c338729f" title="Data collecting method. Returns the mean size of the surface of the environment taken by genotypes en..." alt="" coords="307,325,600,355"/><area shape="rect" id="node17" href="classEcosystem.html#a3af5683a5e2d7c0ff3e8a726f2c5eb32" title="Data collecting method. Returns the standard deviation of the surface of the environment covered by g..." alt="" coords="311,378,596,409"/></map>
</div>
</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="classEcosystem_aa26876007e770024928bd8c2748d2222_icgraph.png" border="0" usemap="#classEcosystem_aa26876007e770024928bd8c2748d2222_icgraph" alt=""/></div>
<map name="classEcosystem_aa26876007e770024928bd8c2748d2222_icgraph" id="classEcosystem_aa26876007e770024928bd8c2748d2222_icgraph">
<area shape="rect" id="node3" href="TheModel_8cpp.html#a0ddf1224851353fc92bfbff6f499fa97" title="The main function." alt="" coords="297,5,348,35"/></map>
</div>
</p>

</div>
</div>
<hr/><h2>Member Data Documentation</h2>
<a class="anchor" id="a46e0a5d150b201ba3fe35bf5087acd07"></a><!-- doxytag: member="Ecosystem::EnvFreeResources" ref="a46e0a5d150b201ba3fe35bf5087acd07" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a46e0a5d150b201ba3fe35bf5087acd07">Ecosystem::EnvFreeResources</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00056">56</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="a52801d9d0de539460a718b0c50eb9e32"></a><!-- doxytag: member="Ecosystem::meanGenomeSize" ref="a52801d9d0de539460a718b0c50eb9e32" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a52801d9d0de539460a718b0c50eb9e32">Ecosystem::meanGenomeSize</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00059">59</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="af66972dddf7dcaa1cabe14e78cb332d1"></a><!-- doxytag: member="Ecosystem::numberOfBirths" ref="af66972dddf7dcaa1cabe14e78cb332d1" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long <a class="el" href="classEcosystem.html#af66972dddf7dcaa1cabe14e78cb332d1">Ecosystem::numberOfBirths</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00062">62</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="abe0e871783bbbda6dd53a6c938f9d02a"></a><!-- doxytag: member="Ecosystem::numberOfDeaths" ref="abe0e871783bbbda6dd53a6c938f9d02a" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long <a class="el" href="classEcosystem.html#abe0e871783bbbda6dd53a6c938f9d02a">Ecosystem::numberOfDeaths</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00061">61</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="a09eb88df55501df846dc482ef7922401"></a><!-- doxytag: member="Ecosystem::PopulationOfCells" ref="a09eb88df55501df846dc482ef7922401" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="classCell.html">Cell</a>&gt; <a class="el" href="classEcosystem.html#a09eb88df55501df846dc482ef7922401">Ecosystem::PopulationOfCells</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00055">55</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="a972b9932e42c702002c7ffc208e009b5"></a><!-- doxytag: member="Ecosystem::startOfMutationCounter" ref="a972b9932e42c702002c7ffc208e009b5" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned long <a class="el" href="classEcosystem.html#a972b9932e42c702002c7ffc208e009b5">Ecosystem::startOfMutationCounter</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00063">63</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<a class="anchor" id="a628b26f0fc25e8d334ba3d9a96dc43eb"></a><!-- doxytag: member="Ecosystem::STD_GenomeSize" ref="a628b26f0fc25e8d334ba3d9a96dc43eb" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double <a class="el" href="classEcosystem.html#a628b26f0fc25e8d334ba3d9a96dc43eb">Ecosystem::STD_GenomeSize</a></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>Definition at line <a class="el" href="ecosystem_8h_source.html#l00060">60</a> of file <a class="el" href="ecosystem_8h_source.html">ecosystem.h</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>00_HGT_CurrentCode/<a class="el" href="ecosystem_8h_source.html">ecosystem.h</a></li>
<li>00_HGT_CurrentCode/<a class="el" href="ecosystem_8cpp_source.html">ecosystem.cpp</a></li>
</ul>
</div><!-- contents -->
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Defines</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="classEcosystem.html">Ecosystem</a>      </li>

    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.7.6.1 </li>
   </ul>
 </div>


</body>
</html>
