<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.7"/>
<title>Genome Streamlining: &quot;Modelling Genome Size Constrains&quot;</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">Genome Streamlining
   &#160;<span id="projectnumber">First Generation Model</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.7 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('index.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Macros</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">"Modelling Genome Size Constrains" </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><dl class="section author"><dt>Author</dt><dd><b>Piotr Bentkowski</b>: <br />
 <a href="#" onclick="location.href='mai'+'lto:'+'ben'+'tk'+'ows'+'ki'+'.pi'+'ot'+'r@g'+'ma'+'il.'+'co'+'m
 '+'?s'+'ubj'+'ec'+'t=G'+'en'+'ome'+' S'+'tre'+'am'+'lin'+'in'+'g M'+'od'+'el'; return false;">bentkowski.piotr@gmail.com</a> <br />
 <br />
 Developed during Piotr's studies at:<br />
 Laboratory for Global Marine and Atmospheric Chemistry <br />
 School of Environmental Sciences <br />
 University of East Anglia <br />
 NR4 7TJ, Norwich <br />
 United Kingdom <br />
 <br />
 Under supervision of Dr. Thomas Mock <br />
 With help from: Dr. Cock van Oosterhout, Dr. Hywel Williams, Prof. Tim Lenton <br />
 </dd></dl>
<dl class="section version"><dt>Version</dt><dd>1.1</dd></dl>
<h1><a class="anchor" id="intro"></a>
Introduction</h1>
<p>This program is an implementation of a model of the evolution of genome size described in Piotr Bentkowski's Ph.D. thesis:<br />
 <a href="https://ueaeprints.uea.ac.uk/50553/">Modelling evolution of genome size in prokaryotes in response to changes in their abiotic environment</a> (2014) defended at the University of East Anglia. <br />
 </p>
<h1><a class="anchor" id="Licence"></a>
Licence</h1>
<p>This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version. <br />
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. <br />
 You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.</p>
<h1><a class="anchor" id="Compilation"></a>
Compilation</h1>
<p>The program has some dependencies on <a href="http://www.gnu.org/software/gsl/">GNU Scientific Library</a>. Should compile smoothly on most GNU/Linux distros with GSL installed. Compilation works fine on Ubuntu 12.04 LTS and 14:04 LTS with a command: <br />
 <br />
 <b> g++ -O2 -o TheModel -L/usr/local/lib <a class="el" href="TheModel_8cpp.html">TheModel.cpp</a> <a class="el" href="genotype_8cpp.html">genotype.cpp</a> <a class="el" href="gene_8cpp.html">gene.cpp</a> <a class="el" href="cell_8cpp.html">cell.cpp</a> <a class="el" href="ecosystem_8cpp.html">ecosystem.cpp</a> <a class="el" href="rngEngine_8cpp.html">rngEngine.cpp</a> <a class="el" href="tagging__system_8cpp.html">tagging_system.cpp</a> -lgsl -lgslcblas -lm </b><br />
 <br />
 It also worked fine on UEA's GRACE Computational Cluster compiled with Intel C++ Compiler (icc)</p>
<h1><a class="anchor" id="Parameters"></a>
Parameters</h1>
<p><b>Program takes exactly 45 parameters. These are:</b> <br />
 <b> 00 </b>- program's name <br />
 <b> 01 </b>- the turbulence level (the length of the step in the random walk; between 0 and 1) <br />
 <b> 02 </b>- resolution of the environment (how dense the environment space is being sampled, well &lt; 1) <br />
 <b> 03 </b>- the value = 1/[env space surface] (calculate it on a sheet of paper cos it's gonna be faster that way) <br />
 <b> 04 </b>- a seed for the Mersenne Twister random number generator (a positive integer) <br />
 <b> 05 </b>- the number of the iterations in the model's run (a positive integer) <br />
 <b> 06 </b>- the 'not-super-gene' parameter (surface under the Gaussian curve)<br />
 <b> 07 </b>- the probability of mutation of existing gene (between 0 and 1) <br />
 <b> 08 </b>- the fraction of env space is occupied by the gene space (suggested value = 1) <br />
 <b> 09 </b>- the probability of deleting a gene (between 0 and 1) <br />
 <b> 10 </b>- the probability of duplicating a gene (between 0 and 1) <br />
 <b> 11 </b>- the maximal number of genes at model's initialisation (don't aim to high, rather dozens then hundreds) <br />
 <b> 12 </b>- the minimal number of genes at initialisation (must be lower then #11)<br />
 <b> 13 </b>- the minimal number of genes at all times (at least 1 must be present or it may result in a segmentation fault) <br />
 <b> 14 </b>- the number of core genes (might be 0)<br />
 <b> 15 </b>- the cost of maintaining one gene (&gt;= 0)<br />
 <b> 16 </b>- the gain from maximal fitness (&gt;= 0 , maximal amount of the resource a cell can gain in one go) <br />
 <b> 17 </b>- the constant metabolic cost (&gt;= 0, a cost of being a cell and being alive)<br />
 <b> 18 </b>- the minimal resource below which a cell dies<br />
 <b> 19 </b>- the minimal resource above which a cell divides<br />
 <b> 20 </b>- the maximal resource bonus given to a cell at the initialisation of the model (actual value is being randomly selected between 0 and this number) <br />
 <b> 21 </b>- the probability of a random death (between 0 and 1)<br />
 <b> 22 </b>- the total amount of resource the ecosystem has (a pretty big number, please)<br />
 <b> 23 </b>- the number of bins in the genome size histogram <br />
 <b> 24 </b>- the number of bins in the histogram of the envelope of genome size <br />
 <b> 25 </b>- the width of the bin in the histogram of the envelope of genome size <br />
 <b> 26 </b>- the number of bins in the histogram of maximums of the fitness function <br />
 <b> 27 </b>- the width of bin in the histogram maximums of the fitness function <br />
 <b> 28 </b>- the number of bins in the histogram of the age of cells <br />
 <b> 29 </b>- the number of bins in the histogram of the resource allocated to cells<br />
 <b> 30 </b>- the width of bins in the histogram of the resource allocated to cells <br />
 <b> 31 </b>- the beginning value of the histogram of the resource allocated to cells <br />
 <b> 32 </b>- the number of bins in the histogram of the resource up-taken by cells in population <br />
 <b> 33 </b>- the width of bins in the histogram of the resource up-taken by cells in population<br />
 <b> 34 </b>- the mode of feeding ('s' -sequential, 'p' -proportional; other values will cause the program to prompt an error message and quit)<br />
 <b> 35 </b>- the parameter how uneven is the resource distribution when a cell is dividing (has meaning only when program works with proportional feeding mode; value between 0 and 0.5) <br />
 <b> 36 </b>- the number of bins in the histogram of the age at the first reproduction <br />
 <b> 37 </b>- the width of bins in the histogram of the age at the first reproduction <br />
 <b> 38 </b>- the number of bins in the histogram of the number of the reproduction events per cell <br />
 <b> 39 </b>- the width of bins in the histogram of the number of the reproduction events per cell <br />
 <b> 40 </b>- the number of bins of the histogram of the average intake per unit of time until reproduction <br />
 <b> 41 </b>- the bin width of the histogram of the average intake per unit of time until reproduction <br />
 <b> 42 </b>- probability that a cell will become a gene donor in HGT <br />
 <b> 43 </b>- probability that a gene will get transferred from the donor's genotype to recipient's genotype <br />
 <b> 44 </b>- time interval (steps) in which statistics are saved to data files<br />
 <br />
 This program can recognise simple errors in the argument list (e.g. value out of range, negative values when positive are needed etc.), but will not recognise when they don't make a 'biological' sense.</p>
<h1><a class="anchor" id="The"></a>
Output</h1>
<p>Program produces a number of text files containing desired data. The file <em>ModelParams.dat</em> contains the values of parameters fed to the model. The file <em>GenomeSizeData.dat</em> contains the main statistics of the model, including the time stamp used in the visualisation process of the other files with data. This two files are the basics to analyse simulation output. Other files are rather optional and they can be switched off by commenting out some lines in the main function (<em><a class="el" href="TheModel_8cpp.html">TheModel.cpp</a></em> file) what should speed up the program runtime. <br />
<br />
 Important files:<br />
 <b><em>ModelParams.dat</em></b> - contains run's parametrisation <br />
 <b><em>GeneralData.dat</em></b> - contains basic statistics of the output <br />
 <br />
 Each line of these files represents one time step (in sync with <em>GeneralData.dat</em>): <br />
 <b><em>GenomeSizeData.dat</em></b> - number of genes in genomes; histograms <br />
 <b><em>GeneNumberOfRepr.dat</em></b> - number of genes in genomes but only of cells which reproduced; histograms <br />
 <b><em>CellsAgeData.dat</em></b> - age of the cells in population; histograms <br />
 <b><em>AgeTillReproduction.dat</em></b> - time passed since the last time cell reproduced (or was born); histograms <br />
 <b><em>FrameSizeData.dat</em></b> - size of the genotype's frame; histograms <br />
 <b><em>FrameMaxData.dat</em></b> - maximums of the genotype's frame; histograms <br />
 <b><em>AvarGainAtRepr.dat</em></b> - cells' average total gain before reproduction; histograms <br />
 <b><em>AvarIntakeAtRepr.dat</em></b> - cells' average intake before reproduction; histograms <br />
 <b><em>RecourceInCells.dat</em></b> - resources allocated into cells in one time step; histograms <br />
 <b><em>NumberOfReproductions.dat</em></b> - how many times a cell reproduced; histograms <br />
 <b><em>RecourceUptakenCells.dat</em></b> - resources taken by cells in one time step; histograms <br />
 <b><em>AvaregeGenotype.dat</em></b> - U(x) values (intake values) in all the cells in the population; histograms <br />
 <b><em>AvarIntakeAtRepr.dat</em></b> - cells' average intake before reproduction; histograms <br />
 <b><em>FrameMaxData.dat</em></b> - maximums of the genotypes' frame; histograms <br />
 <b><em>GeneNumberOfRepr.dat</em></b> - cells' gene number distribution, only cells which reproduced; histograms <br />
 <b><em>GenotypeEnvelMean.dat</em></b> - mean values of U(x) function (the genome's envelope) <br />
 <b><em>GenotypeEnvelSTD.dat</em></b> - SD of U(x) function (the genome's envelope) <br />
 <br />
 Other data files: <br />
 <b><em>MutationRecords.dat</em></b> - cumulative number of mutations accumulated by each living cell in a defined time period. There is a dedicated function <a class="el" href="classEcosystem.html#a4227871d6d47201edf4450fd35882c53" title="Data collecting method. Clears all three record of all the cells regarding the number of mutations...">Ecosystem::clearMutationCounters(int current_iteration)</a> for zeroing number of mutation <br />
 <b><em>GenomesOfPopulation.dat</em></b> - these are all the genomes in the ecosystem <br />
 </p>
<h1><a class="anchor" id="Visualisation"></a>
Visualisation</h1>
<p>Visualisation is done using Python scripts containing calls to Pylab library (Linux distros will have it in its repos, for Windows you might check the <a href="http://code.google.com/p/pythonxy/">Python(x,y) project</a>). These scripts can be found in <em>01_PyScripts</em> directory. </p>
</div></div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Thu Jan 28 2016 16:59:51 for Genome Streamlining by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.7 </li>
  </ul>
</div>
</body>
</html>
